File: 7/4/nesting.R
# Determine whether a given string of parentheses (single type) is properly nested.
nesting <- function(str_s){
  # set the balance of parentheses at zero
  balance <- 0
  # convert string to vector
  str_s <- strsplit(str_s, "")[[1]]
  # iterate through the vector of parentheses
  for(paren in str_s){
    # if parenthesis is left, add to balance; otherwise, subtract
    ifelse(paren == "(", balance <- balance + 1, balance <- balance - 1)
      # there should never be more right parentheses
    if(balance < 0){
      return(0)
    }
  }
  # a balance of zero indecates balance
  return(ifelse(balance == 0, 1, 0))
}
# Debugging:
# print(nesting("())"))
# print(nesting("(()(())())"))
# Bash Testing:
ARGS_S <- commandArgs()[6]
print(nesting(ARGS_S))

File: 7/4/nesting.py
"""Determine whether a given string of parentheses (single type) is properly nested."""
from __future__ import print_function
import getopt
import sys
def nesting(str_s):
    """Determine whether a given string of parentheses (single type) is properly nested."""
    # set the balance of parentheses at zero
    balance = 0
    # iterate through the string of parentheses
    for paren in str_s:
        # test whether it's a left parenthesis
        if paren == "(":
            # add to balance for left parentheses
            balance += 1
        else:
            # subtract from balance for right parentheses
            balance -= 1
            # there should never be more right parentheses
            if balance < 0:
                # if there are, return zero
                return 0
    # a balance of zero indicates balance
    if balance == 0:
        return 1
    # any other balance score indicates imbalance
    return 0
# Debugging:
# nesting('())')
# nesting('(()(())())')
# Bash Testing:
ARGS_S = list(getopt.getopt(sys.argv, "ho:v"))
ARGS_S = ARGS_S[1][1]
print(nesting(ARGS_S))
# Codility Testing:
# https://app.codility.com/demo/results/trainingNYDRZT-QFC/
# Correctness: 100%
# Performance: 100%
# Difficulty: Painless
